File: packages/typescript/src/types/webhooks.ts
import type { TransactionStatus, TransactionType } from './common.js';

export interface WebhookTransaction {
  id: string;
  tenant_id: string;
  type: TransactionType;
  status: TransactionStatus;
  amount: number;
  currency: string;
  phone_number: string;
  external_reference: string | null;
  provider: string;
  provider_request_id: string | null;
  provider_checkout_id: string | null;
  result_code: string | null;
  result_desc: string | null;
  mpesa_receipt: string | null;
  created_at: string;
  updated_at: string;
}

export type WebhookEventType =
  | 'transaction.success'
  | 'transaction.failed'
  | 'transaction.cancelled'
  | 'transaction.expired'
  | 'transaction.updated';

export interface WebhookPayload {
  event: 'transaction.updated';
  event_type: WebhookEventType;
  transaction: WebhookTransaction;
}

File: packages/python/palpluss/types/webhooks.py
from __future__ import annotations

from typing import Literal

from typing_extensions import TypedDict

from .common import TransactionStatus, TransactionType

WebhookEventType = Literal[
    "transaction.success",
    "transaction.failed",
    "transaction.cancelled",
    "transaction.expired",
    "transaction.updated",
]


class WebhookTransaction(TypedDict):
    id: str
    tenant_id: str
    type: TransactionType
    status: TransactionStatus
    amount: float
    currency: str
    phone_number: str
    external_reference: str | None
    provider: str
    provider_request_id: str | None
    provider_checkout_id: str | None
    result_code: str | None
    result_desc: str | None
    mpesa_receipt: str | None
    created_at: str
    updated_at: str


class WebhookPayload(TypedDict):
    event: str
    event_type: WebhookEventType
    transaction: WebhookTransaction
